tools/topology/topology2/CMakeLists.txt

```
@@ -3,10 +3,16 @@ add_custom_target(topologies2
 	COMMAND ${CMAKE_COMMAND} -E copy_directory cavs ${SOF_TOPOLOGY_BINARY_DIRECTORY}
 )
 
-# generate ABI
+# generate ABI for IPC3
 execute_process(
-	COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/get_abi.sh ${SOF_ROOT_SOURCE_DIRECTORY}
-	OUTPUT_FILE ${CMAKE_CURRENT_BINARY_DIR}/abi.conf
+	COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/get_abi_ipc3.sh ${SOF_ROOT_SOURCE_DIRECTORY}
+	OUTPUT_FILE ${CMAKE_CURRENT_BINARY_DIR}/abi_ipc3.conf
+)
+
+# generate ABI for IPC4
+execute_process(
+	COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/get_abi_ipc4.sh ${SOF_ROOT_SOURCE_DIRECTORY}
```

Come on, more than half of the new IPC4 script is a straight copy/paste of the other one... I doubt you would do this in C!

```suggestion
	COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/get_abi.sh u16 (or u8) ${SOF_ROOT_SOURCE_DIRECTORY}
```
Then in a single, common `get_abi.sh` script:

```shell

for vers in $ABI_MAJOR $ABI_MINOR do

  case "$1" in
     u8) print_one_byte "$vers" ;;
     u16) print_two_bytes "$vers" ;;
  esac

done

```
with `print_two_bytes` as defined in an earlier comment.

---

and where will I pass this u16 or u8 to cmake from?

---

the point is that I need to know if I am building a topology for IPC4 or IPC3 before I execute this script and it could be different for each file. I could move the execute process inside the cavs/CMakeLists.txt if you want to avoid 2 scripts

---

ok i moved it to the cavs CMakeLists.txt now. better?